### PR TITLE
Refining error message text

### PR DESCRIPTION
**Pull request by sonalmahajan15**
This PR refines the error message text of produce and consume triggers in order to improve the overall UX.

[Depends on PR #31]

### CODE REVIEW

File: config/const.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ const StableRoundLimit = 5
 // BackpropTimeout is the timeout set for analysis of each function. This ensures build time SLAs.
 // NilAway should report an error if this timeout is ever hit, in order not to silently ignore any
 // functions due to this.
-const BackpropTimeout = 10 * time.Second
+const BackpropTimeout = 10000 * time.Second
```
Comment (lazaroclapp) [inline config/const.go:22]: Why are we increasing the timeout here to multiple hours? Leftover debugging change?

File: inference/conflict.go [hunk 2]
```diff
@@ -97,7 +97,8 @@ func (n *nilFlow) String() string {
 	lastNonnilNodeIndex := len(n.nonnilPath) - 1
 	n.nonnilPath[lastNonnilNodeIndex].reason = fmt.Sprintf("%s (must be NONNIL)", n.nonnilPath[lastNonnilNodeIndex].reason)
 
-	return pathString(n.nilPath) + pathString(n.nonnilPath)
+	flow := pathString(n.nilPath) + pathString(n.nonnilPath)
+	return flow
```
Comment (lazaroclapp) [inline inference/conflict.go:101]: Why this change?

File: util/util.go [hunk 3]
```diff
@@ -452,7 +452,8 @@ func ExprToString(e ast.Expr, pass *analysis.Pass) string {
 	var buf bytes.Buffer
 	err := printer.Fprint(&buf, pass.Fset, e)
 	if err != nil {
-		panic(fmt.Sprintf("Failed to convert AST expression to string: %v\n", err))
+		// printer failed to convert AST expression to string, hence return empty string to avoid any panics downstream
+		return ""
```
Comment (lazaroclapp) [inline util/util.go:456]: When is this expected to happen?

File: config/const.go (hunk 1 above)
Comment (sonalmahajan15) [reply config/const.go:22]: Yes 😄. Reverted.

File: inference/conflict.go (hunk 2 above)
Comment (sonalmahajan15) [reply inference/conflict.go]: It seems to be stemming from a rebasing issue.

File: annotation/consume_trigger.go [hunk 4]
```diff
@@ -204,6 +204,8 @@ func (SliceAccessPrestring) String() string {
 // FldAccess is when a value flows to a point where a field of it is accessed, and so it must be non-nil
 type FldAccess struct {
 	ConsumeTriggerTautology
+	Field  *types.Var
+	Method *types.Func
```
Comment (yuxincs) [inline annotation/consume_trigger.go:208]: IIUC, one and only one of these two must be nonnil, right? Can we make that explicit in the comments here?

Alternatively, we can just store `Sel types.Object` here and use type assertions to determine whether it is `*types.Var` or `*types.Func`

File: annotation/consume_trigger.go [hunk 5]
```diff
@@ -212,14 +214,30 @@ func (f FldAccess) String() string {
 
 // Prestring returns this FldAccess as a Prestring
 func (f FldAccess) Prestring() Prestring {
-	return FldAccessPrestring{}
+	fieldName, methodName := "", ""
+	if f.Field != nil {
+		fieldName = f.Field.Name()
+	} else if f.Method != nil {
+		methodName = f.Method.Name()
+	}
```
Comment (yuxincs) [inline annotation/consume_trigger.go:222]: panic otherwise?

File: inference/conflict.go [hunk 6]
```diff
@@ -110,18 +110,18 @@ func pathString(nodes []node) string {
...
+	allNodes = append(allNodes, n.nilPath...)
+	allNodes = append(allNodes, n.nonnilPath...)
+
 	flow := ""
-	for _, nodes := range [...][]node{n.nilPath, n.nonnilPath} {
-		for _, nodeObj := range nodes {
-			flow += nodeObj.String()
-		}
+	for _, nodeObj := range allNodes {
+		flow += nodeObj.String()
```
Comment (yuxincs) [inline inference/conflict.go:119]: Any reason for this change?

File: inference/conflict.go [hunk 7]
```diff
@@ -69,16 +69,16 @@ func newNode(p annotation.Prestring, c annotation.Prestring) node {
...
-	} else if len(n.producerRepr) > 0 {
-		reasonStr = n.producerRepr
-	} else if len(n.consumerRepr) > 0 {
-		reasonStr = n.consumerRepr
+
+	if len(n.producerRepr) > 0 {
+		reasonStr += n.producerRepr + " "
+	}
+	if len(n.consumerRepr) > 0 {
+		reasonStr += n.consumerRepr
```
Comment (yuxincs) [inline inference/conflict.go:81]: So this means if `n.producerRepr` is non-empty and `n.consumerRepr` is empty, we would result in a string `<producerRepr> ` (notice the trailing space).

Not that it matters much from UX perspective, but perhaps we do not want that anyways?

File: annotation/produce_trigger.go [hunk 8]
```diff
@@ -333,28 +333,33 @@ func (UnassignedFld) Prestring() Prestring {
...
 
 func (UnassignedFldPrestring) String() string {
-	return "unassigned at struct initialization"
+	return "uninitialized"
 }
 
 // NoVarAssign is when a value is determined to flow from a variable that wasn't assigned to
 type NoVarAssign struct {
 	ProduceTriggerTautology
+	VarDecl *types.Var
```
Comment (yuxincs) [inline annotation/produce_trigger.go:342]:
```suggestion
	VarObj *types.Var
```

I know this is done for consistency with the existing codebase. But I feel we should use `XXObj` instead of `XXDecl` for `*types.XX` since it refers to the object (entity) but not its declaration (which is an AST node). 

See more at https://github.com/golang/example/blob/master/gotypes/go-types.md

**Review by yuxincs — APPROVED** [review]
Great error message refinement!!

I have a few nits and a question, but none of them are blocking

File: annotation/consume_trigger.go (hunk 4 above)
Comment (sonalmahajan15) [reply annotation/consume_trigger.go:208]: Good idea. Went ahead with the latter option.

File: inference/conflict.go (hunk 6 above)
Comment (sonalmahajan15) [reply inference/conflict.go:119]: Short answer is it was causing errors being dropped.

Long answer: Single assertion conflict has only one producer and consumer. Now with the concatenation logic, this got combined in a single node which is stored in the `nonnilPath`. Thus keeping `nilPath` nil. In such cases , `for _, nodes := range [...][]node{n.nilPath, n.nonnilPath} {` was causing the combined slice to be nil, causing the loop to be skipped, dropping errors.

File: inference/conflict.go (hunk 7 above)
Comment (sonalmahajan15) [reply inference/conflict.go:81]: Good catch. Fixed.

File: annotation/consume_trigger.go (hunk 5 above)
Comment (sonalmahajan15) [reply annotation/consume_trigger.go:222]: Done

File: annotation/produce_trigger.go (hunk 8 above)
Comment (sonalmahajan15) [reply annotation/produce_trigger.go:342]: Right, changed here. But we should systematically change the entire codebase to follow the convention.
